### PR TITLE
New structure for virt-who plugin cases

### PR DESCRIPTION
**Pull request by hkx303**
Hello, this a large PR, I would be very grateful to everyone who involved.
  
**Description**
Recently, we created an issue(https://github.com/SatelliteQE/robottelo/issues/7871) to optimize virt-who cases to the new structure, it would be easy to integrate them to the existing Sat-Jenkins:
```
── test/foreman/virtwho
    └── ui
        ├── test_esx.py
        ├── test_hyperv.py
        ├── test_kubevirt.py
        ├── test_libvirt.py
        ├── test_rhevm.py
        └── test_xen.py
    └── cli
        ├── test_esx.py
        ├── test_hyperv.py
        ├── test_kubevirt.py
        ├── test_libvirt.py
        ├── test_rhevm.py
        └── test_xen.py
    └── api
        ├── test_esx.py
        ├── test_hyperv.py
        ├── test_kubevirt.py
        ├── test_libvirt.py
        ├── test_rhevm.py
        └── test_xen.py
```

Main changes:

- [x] Re-write the way to read the config file for virt-who testing
     With the new structure, we need the config section for 6 different hypervisors and need to control it separately, so we need the new config file for those settings, for details, see (https://github.com/SatelliteQE/robottelo/pull/7919)
      1. Create a new config file for virt-who hypervisors (**virtwho.properties**)
      2. Create the config/virtwho.py to read the settings for the config file
      3. Update all test cases for virt-who plugin based on the new way.
      In the future, if we change this to Dynaconf, we will also update.

- [x] Re-organize all test cases for the virt-who plugin based on the new structure(https://github.com/SatelliteQE/robottelo/pull/7906)
      1. Create test files for each type of hypervisor.
      2. Update the specific `form_data` in the test cases for each type of hypervisor.
      3. Make the cases more prominent for each hypervisor by adding/deleting some cases
      

**Test Result**
We have created the separate branch for these change: [new_struc_for_virtwho_plugin](https://github.com/SatelliteQE/robottelo/tree/new_struc_for_virtwho_plugin), updated our virt-who jenkins job based on the new structure, can see the whole test result in:
https://virtwho-qe-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/view/Virtwho-Configure/job/Virtwho-Plugin-Runtest/5/console
```
pytest /home/jenkins/workspace/Virtwho-Plugin-Runtest/robottelo/tests/foreman/virtwho --disable-pytest-warnings 
-v --junit-xml=/home/jenkins/workspace/Virtwho-Plugin-Runtest/report/results.xml 
-o junit_suite_name=report_xml --html=/home/jenkins/workspace/Virtwho-Plugin-Runtest/report/index.html
============================= test session starts ==============================
platform linux -- Python 3.7.6, pytest-4.6.3, py-1.7.0, pluggy-0.13.1 -- /usr/bin/python3
cachedir: .pytest_cache
shared_function enabled - OFF - scope:  - storage: file
rootdir: /home/jenkins/workspace/Virtwho-Plugin-Runtest/robottelo
plugins: html-2.0.1, metadata-1.8.0, services-1.3.1, mock-1.10.4, cov-2.10.1, flake8-1.0.1, forked-1.0.1, xdist-1.27.0
collecting ... 2020-08-28 20:43:09 - conftest - DEBUG - Collected 78 test cases
collected 78 items

robottelo/tests/foreman/virtwho/api/test_esx.py::TestVirtWhoConfigforEsx::test_positive_deploy_configure_by_id PASSED [  1%]
robottelo/tests/foreman/virtwho/api/test_esx.py::TestVirtWhoConfigforEsx::test_positive_deploy_configure_by_script PASSED [  2%]
robottelo/tests/foreman/virtwho/api/test_esx.py::TestVirtWhoConfigforEsx::test_positive_debug_option PASSED [  3%]
robottelo/tests/foreman/virtwho/api/test_esx.py::TestVirtWhoConfigforEsx::test_positive_interval_option PASSED [  5%]
robottelo/tests/foreman/virtwho/api/test_esx.py::TestVirtWhoConfigforEsx::test_positive_hypervisor_id_option PASSED [  6%]
robottelo/tests/foreman/virtwho/api/test_esx.py::TestVirtWhoConfigforEsx::test_positive_filter_option PASSED [  7%]
robottelo/tests/foreman/virtwho/api/test_esx.py::TestVirtWhoConfigforEsx::test_positive_proxy_option PASSED [  8%]
robottelo/tests/foreman/virtwho/api/test_esx.py::TestVirtWhoConfigforEsx::test_positive_configure_organization_list PASSED [ 10%]
robottelo/tests/foreman/virtwho/api/test_hyperv.py::TestVirtWhoConfigforHyperv::test_positive_deploy_configure_by_id PASSED [ 11%]
robottelo/tests/foreman/virtwho/api/test_hyperv.py::TestVirtWhoConfigforHyperv::test_positive_deploy_configure_by_script PASSED [ 12%]
robottelo/tests/foreman/virtwho/api/test_hyperv.py::TestVirtWhoConfigforHyperv::test_positive_hypervisor_id_option PASSED [ 14%]
robottelo/tests/foreman/virtwho/api/test_kubevirt.py::TestVirtWhoConfigforKubevirt::test_positive_deploy_configure_by_id SKIPPED [ 15%]
robottelo/tests/foreman/virtwho/api/test_kubevirt.py::TestVirtWhoConfigforKubevirt::test_positive_deploy_configure_by_script SKIPPED [ 16%]
robottelo/tests/foreman/virtwho/api/test_kubevirt.py::TestVirtWhoConfigforKubevirt::test_positive_hypervisor_id_option SKIPPED [ 17%]
robottelo/tests/foreman/virtwho/api/test_libvirt.py::TestVirtWhoConfigforLibvirt::test_positive_deploy_configure_by_id PASSED [ 19%]
robottelo/tests/foreman/virtwho/api/test_libvirt.py::TestVirtWhoConfigforLibvirt::test_positive_deploy_configure_by_script PASSED [ 20%]
robottelo/tests/foreman/virtwho/api/test_libvirt.py::TestVirtWhoConfigforLibvirt::test_positive_hypervisor_id_option PASSED [ 21%]
robottelo/tests/foreman/virtwho/api/test_rhevm.py::TestVirtWhoConfigforRhevm::test_positive_deploy_configure_by_id PASSED [ 23%]
robottelo/tests/foreman/virtwho/api/test_rhevm.py::TestVirtWhoConfigforRhevm::test_positive_deploy_configure_by_script PASSED [ 24%]
robottelo/tests/foreman/virtwho/api/test_rhevm.py::TestVirtWhoConfigforRhevm::test_positive_hypervisor_id_option PASSED [ 25%]
robottelo/tests/foreman/virtwho/api/test_xen.py::TestVirtWhoConfigforXen::test_positive_deploy_configure_by_id PASSED [ 26%]
robottelo/tests/foreman/virtwho/api/test_xen.py::TestVirtWhoConfigforXen::test_positive_deploy_configure_by_script PASSED [ 28%]
robottelo/tests/foreman/virtwho/api/test_xen.py::TestVirtWhoConfigforXen::test_positive_hypervisor_id_option PASSED [ 29%]
robottelo/tests/foreman/virtwho/cli/test_esx.py::TestVirtWhoConfigforEsx::test_positive_deploy_configure_by_id PASSED [ 30%]
robottelo/tests/foreman/virtwho/cli/test_esx.py::TestVirtWhoConfigforEsx::test_positive_deploy_configure_by_script PASSED [ 32%]
robottelo/tests/foreman/virtwho/cli/test_esx.py::TestVirtWhoConfigforEsx::test_positive_debug_option PASSED [ 33%]
robottelo/tests/foreman/virtwho/cli/test_esx.py::TestVirtWhoConfigforEsx::test_positive_interval_option PASSED [ 34%]
robottelo/tests/foreman/virtwho/cli/test_esx.py::TestVirtWhoConfigforEsx::test_positive_hypervisor_id_option PASSED [ 35%]
robottelo/tests/foreman/virtwho/cli/test_esx.py::TestVirtWhoConfigforEsx::test_positive_filter_option PASSED [ 37%]
robottelo/tests/foreman/virtwho/cli/test_esx.py::TestVirtWhoConfigforEsx::test_positive_proxy_option PASSED [ 38%]
robottelo/tests/foreman/virtwho/cli/test_esx.py::TestVirtWhoConfigforEsx::test_positive_rhsm_option PASSED [ 39%]
robottelo/tests/foreman/virtwho/cli/test_esx.py::TestVirtWhoConfigforEsx::test_positive_post_hypervisors PASSED [ 41%]
robottelo/tests/foreman/virtwho/cli/test_esx.py::TestVirtWhoConfigforEsx::test_positive_foreman_packages_protection PASSED [ 42%]
robottelo/tests/foreman/virtwho/cli/test_hyperv.py::TestVirtWhoConfigforHyperv::test_positive_deploy_configure_by_id PASSED [ 43%]
robottelo/tests/foreman/virtwho/cli/test_hyperv.py::TestVirtWhoConfigforHyperv::test_positive_deploy_configure_by_script PASSED [ 44%]
robottelo/tests/foreman/virtwho/cli/test_hyperv.py::TestVirtWhoConfigforHyperv::test_positive_hypervisor_id_option PASSED [ 46%]
robottelo/tests/foreman/virtwho/cli/test_kubevirt.py::TestVirtWhoConfigforKubevirt::test_positive_deploy_configure_by_id SKIPPED [ 47%]
robottelo/tests/foreman/virtwho/cli/test_kubevirt.py::TestVirtWhoConfigforKubevirt::test_positive_deploy_configure_by_script SKIPPED [ 48%]
robottelo/tests/foreman/virtwho/cli/test_kubevirt.py::TestVirtWhoConfigforKubevirt::test_positive_hypervisor_id_option SKIPPED [ 50%]
robottelo/tests/foreman/virtwho/cli/test_libvirt.py::TestVirtWhoConfigforLibvirt::test_positive_deploy_configure_by_id PASSED [ 51%]
robottelo/tests/foreman/virtwho/cli/test_libvirt.py::TestVirtWhoConfigforLibvirt::test_positive_deploy_configure_by_script PASSED [ 52%]
robottelo/tests/foreman/virtwho/cli/test_libvirt.py::TestVirtWhoConfigforLibvirt::test_positive_hypervisor_id_option PASSED [ 53%]
robottelo/tests/foreman/virtwho/cli/test_rhevm.py::TestVirtWhoConfigforRhevm::test_positive_deploy_configure_by_id PASSED [ 55%]
robottelo/tests/foreman/virtwho/cli/test_rhevm.py::TestVirtWhoConfigforRhevm::test_positive_deploy_configure_by_script PASSED [ 56%]
robottelo/tests/foreman/virtwho/cli/test_rhevm.py::TestVirtWhoConfigforRhevm::test_positive_hypervisor_id_option PASSED [ 57%]
robottelo/tests/foreman/virtwho/cli/test_xen.py::TestVirtWhoConfigforXen::test_positive_deploy_configure_by_id PASSED [ 58%]
robottelo/tests/foreman/virtwho/cli/test_xen.py::TestVirtWhoConfigforXen::test_positive_deploy_configure_by_script PASSED [ 60%]
robottelo/tests/foreman/virtwho/cli/test_xen.py::TestVirtWhoConfigforXen::test_positive_hypervisor_id_option PASSED [ 61%]
robottelo/tests/foreman/virtwho/ui/test_esx.py::TestVirtwhoConfigforEsx::test_positive_deploy_configure_by_id PASSED [ 62%]
robottelo/tests/foreman/virtwho/ui/test_esx.py::TestVirtwhoConfigforEsx::test_positive_deploy_configure_by_script PASSED [ 64%]
robottelo/tests/foreman/virtwho/ui/test_esx.py::TestVirtwhoConfigforEsx::test_positive_debug_option PASSED [ 65%]
robottelo/tests/foreman/virtwho/ui/test_esx.py::TestVirtwhoConfigforEsx::test_positive_interval_option PASSED [ 66%]
robottelo/tests/foreman/virtwho/ui/test_esx.py::TestVirtwhoConfigforEsx::test_positive_hypervisor_id_option PASSED [ 67%]
robottelo/tests/foreman/virtwho/ui/test_esx.py::TestVirtwhoConfigforEsx::test_positive_filtering_option PASSED [ 69%]
robottelo/tests/foreman/virtwho/ui/test_esx.py::TestVirtwhoConfigforEsx::test_positive_proxy_option PASSED [ 70%]
robottelo/tests/foreman/virtwho/ui/test_esx.py::TestVirtwhoConfigforEsx::test_positive_virtwho_roles PASSED [ 71%]
robottelo/tests/foreman/virtwho/ui/test_esx.py::TestVirtwhoConfigforEsx::test_positive_virtwho_configs_widget PASSED [ 73%]
robottelo/tests/foreman/virtwho/ui/test_esx.py::TestVirtwhoConfigforEsx::test_positive_delete_configure PASSED [ 74%]
robottelo/tests/foreman/virtwho/ui/test_esx.py::TestVirtwhoConfigforEsx::test_positive_virtwho_reporter_role PASSED [ 75%]
robottelo/tests/foreman/virtwho/ui/test_esx.py::TestVirtwhoConfigforEsx::test_positive_virtwho_viewer_role PASSED [ 76%]
robottelo/tests/foreman/virtwho/ui/test_esx.py::TestVirtwhoConfigforEsx::test_positive_virtwho_manager_role PASSED [ 78%]
robottelo/tests/foreman/virtwho/ui/test_esx.py::TestVirtwhoConfigforEsx::test_positive_overview_label_name PASSED [ 79%]
robottelo/tests/foreman/virtwho/ui/test_esx.py::TestVirtwhoConfigforEsx::test_positive_last_checkin_status PASSED [ 80%]
robottelo/tests/foreman/virtwho/ui/test_hyperv.py::TestVirtwhoConfigforHyperv::test_positive_deploy_configure_by_id PASSED [ 82%]
robottelo/tests/foreman/virtwho/ui/test_hyperv.py::TestVirtwhoConfigforHyperv::test_positive_deploy_configure_by_script PASSED [ 83%]
robottelo/tests/foreman/virtwho/ui/test_hyperv.py::TestVirtwhoConfigforHyperv::test_positive_hypervisor_id_option PASSED [ 84%]
robottelo/tests/foreman/virtwho/ui/test_kubevirt.py::TestVirtwhoConfigforKubevirt::test_positive_deploy_configure_by_id PASSED [ 85%]
robottelo/tests/foreman/virtwho/ui/test_kubevirt.py::TestVirtwhoConfigforKubevirt::test_positive_deploy_configure_by_script PASSED [ 87%]
robottelo/tests/foreman/virtwho/ui/test_kubevirt.py::TestVirtwhoConfigforKubevirt::test_positive_hypervisor_id_option PASSED [ 88%]
robottelo/tests/foreman/virtwho/ui/test_libvirt.py::TestVirtwhoConfigforLibvirt::test_positive_deploy_configure_by_id PASSED [ 89%]
robottelo/tests/foreman/virtwho/ui/test_libvirt.py::TestVirtwhoConfigforLibvirt::test_positive_deploy_configure_by_script PASSED [ 91%]
robottelo/tests/foreman/virtwho/ui/test_libvirt.py::TestVirtwhoConfigforLibvirt::test_positive_hypervisor_id_option PASSED [ 92%]
robottelo/tests/foreman/virtwho/ui/test_rhevm.py::TestVirtwhoConfigforRhevm::test_positive_deploy_configure_by_id PASSED [ 93%]
robottelo/tests/foreman/virtwho/ui/test_rhevm.py::TestVirtwhoConfigforRhevm::test_positive_deploy_configure_by_script PASSED [ 94%]
robottelo/tests/foreman/virtwho/ui/test_rhevm.py::TestVirtwhoConfigforRhevm::test_positive_hypervisor_id_option PASSED [ 96%]
robottelo/tests/foreman/virtwho/ui/test_xen.py::TestVirtwhoConfigforXen::test_positive_deploy_configure_by_id PASSED [ 97%]
robottelo/tests/foreman/virtwho/ui/test_xen.py::TestVirtwhoConfigforXen::test_positive_deploy_configure_by_script PASSED [ 98%]
robottelo/tests/foreman/virtwho/ui/test_xen.py::TestVirtwhoConfigforXen::test_positive_hypervisor_id_option PASSED [100%]

 generated xml file: /home/jenkins/workspace/Virtwho-Plugin-Runtest/report/results.xml 
 generated html file: file:///home/jenkins/workspace/Virtwho-Plugin-Runtest/report/index.html 
============ 72 passed, 6 skipped, 92 warnings in 20413.19 seconds =============
```

### CODE REVIEW

**Review by eko999 — APPROVED** [review]
This is really a big improvement for virt-who plugin:
1. All the hypervisors configuration are managed by a separated file: virtwho.properties, next we will push this file to satellite jenkins repository, and then virt-who plugin cases can be run by satellite jenkins easily.
2. The test cases are re-organized by different hypervisors, the benefit is:

- help us to understand the specific test cases for different hypervisor and easy management

- help us to debug and analyse the issue for different hypervisor easily

- reduce the test time, we need almost 17 hours to run the cases in the old structure, but now we only 6-7 hours to run all the cases.

File: robottelo/config/virtwho.py
Comment (JacobCallahan) [inline robottelo/config/virtwho.py:94]: same question as above

**Review by JacobCallahan — CHANGES_REQUESTED** [review]

File: robottelo/config/virtwho.py
Comment (hkx303) [reply robottelo/config/virtwho.py:31]: Yeah, copied them from base.py, and then I added some contents, forgot to delete them. Thanks, will update this with the next commit :)

File: robottelo/config/virtwho.py
Comment (hkx303) [reply robottelo/config/virtwho.py:94]: Imported FeatureSettings from the base.py file

File: robottelo/config/virtwho.py
Comment (hkx303) [reply robottelo/config/virtwho.py:28]: also import get_project_root from base.py to avoid the repetitive functions.

**Comment by hkx303** [comment]
Thank you @JacobCallahan  for the review as this is really a large PR. I have imported those functions from base.py to avoid repetitive functions.

**Comment by hkx303** [comment]
Hey @JacobCallahan , Could you take a look at this again, thanks

File: robottelo/virtwho_utils.py
Comment (mirekdlugosz) [inline robottelo/virtwho_utils.py:45]: `eval` is usually highly discouraged. You can use `getattr` here instead:

```
'hostname': getattr(virtwho, system_type).guest,
```

File: robottelo/config/virtwho.py
Comment (mirekdlugosz) [inline robottelo/config/virtwho.py:8]: Just for your information, there is plan to move to Dynaconf, which will store configuration in YAML files. See #8003 . At some unspecified time in the future, `robottelo.config.base` will be removed.

I don't want to block this PR, but I want to give you heads up and opportunity to re-think how these configurations will be stored. YAML gives you opportunity to store values in multilevel namespaces, make values proper complex data structures (like lists or dictionaries) or re-use part of configuration in many places.

**Review by mirekdlugosz — CHANGES_REQUESTED** [review]
I would really like us to not use `eval`. This is only thing I consider blocking.

I'm also not completely sold on idea of repeating basically the same tests in multiple files, one for each hypervisor. I know there are some differences between hypervisors, but on the other hand, there are also many similarities. From my point of view, using pytest parameters would be better idea.

But I assume you guys have given proper thought to this already, and this PR has been waiting for review long enough, so I don't want to put additional obstacles for you.

Other than that, great job!

File: robottelo/config/virtwho.py
Comment (hkx303) [reply robottelo/config/virtwho.py:8]: Yes, I saw your PR when it was raised cuz I also focus on the Dynaconf integration job, and really happy with your fantastic progress.

We did think about how to set our configs, and the job move to Dynaconf is still in progress,  so just as I said in the description of this PR, we will also update our configurations once the Dynaconf PR is merged/the `robottelo.config.base` is removed. We will sync up with the SATQE team, do you think this will be ok? Anyway, thanks for bringing up this.

File: robottelo/virtwho_utils.py
Comment (hkx303) [reply robottelo/virtwho_utils.py:45]: OK, will update this with the next commit.

Comment (hkx303) [reply robottelo/virtwho_utils.py]: and is it because `it is slow, it is prone to bugs and strange behavior, and it makes the code hard to handle and very difficult to understand. `？

Comment (mirekdlugosz) [reply robottelo/virtwho_utils.py]: @hkx303 all of these reasons, plus eval is potentially dangerous and might be root cause of security breach.
Basically, it's best to avoid eval unless you absolutely have to. And here you don't have to.

File: robottelo/config/virtwho.py
Comment (mirekdlugosz) [reply robottelo/config/virtwho.py:8]: Yes, syncing with dynaconf config later on is OK. Dynaconf PR is not blocking this one, but I wanted to ensure you are aware of it and that you plan for it accordingly.

**Review by mirekdlugosz — APPROVED** [review]
My main concern is resolved.

Waiting for @JacobCallahan , who NACK'ed this earlier. If all his comments are resolved, I think we can finally merge this.

File: robottelo/config/virtwho.py
Comment (hkx303) [reply robottelo/config/virtwho.py:8]: Gotcha, thanks

**Review by JacobCallahan — APPROVED** [review]
ACK, great addition!

**Comment by jyejare** [comment]
Not sure why we have not merged this PR yet. We have two tier2 Acks.

**Comment by eko999** [comment]
Hey all, I'd like to merge this PR now since it has 3 approvals, in addition, we can start our next plans without pendings.

**Comment by hkx303** [comment]
Thank you all, really appreciate it!!

**Pull request merged**
